src/xmerl_dsig.erl

```diff
@@ -0,0 +1,730 @@
+    <<"<?xml version=\"1.0\"?>", Data/binary>> =
+        unicode:characters_to_binary(SigInfoXml, unicode, utf8),
+    logger:debug("SigInfoCanonData:~n~p~n", [Data]),
+
+    Sig = public_key:sign(Data, HashFunction, PrivateKey),
+    Sig64 = base64:encode_to_string(Sig),
+    logger:debug("CertBin: ~p~n", [CertBin]),
+    Cert64 = base64:encode_to_string(CertBin),
+    SubjectName = subject_name(CertBin),
+
+    %% and wrap it all up with the signature and certificate
+    SigValue = get_child(DsPrefix, "SignatureValue", Signature),
+    NewSigValue = SigValue#xmlElement{
+                    content = [#xmlText{value = Sig64}]},
+    Sig0 = replace_child(NewSigValue, Signature),
+    Sig1 = replace_child(NewSigInfo, Sig0),
+    KeyInfo = get_child(DsPrefix, "KeyInfo", Signature),
+    X509Data = get_child(DsPrefix, "X509Data", KeyInfo),
+    X509SubjectName = get_child(DsPrefix, "X509SubjectName", X509Data),
+    NewX509SubjectName =
+        X509SubjectName#xmlElement{
+          content = [#xmlText{value = SubjectName}]},
+    X509Certificate = get_child(DsPrefix, "X509Certificate", X509Data),
+    NewX509Certificate =
+        X509Certificate#xmlElement{
+          content = [#xmlText{value = Cert64}]},
+    NewX509Data = replace_child(NewX509Certificate, X509Data),
+    FinalX509Data = replace_child(NewX509SubjectName, NewX509Data),
+    NewKeyInfo = replace_child(FinalX509Data, KeyInfo),
+    FinalSig = replace_child(NewKeyInfo, Sig1),
+    ElementOut = dreplace_child(FinalSig, ElementIn),
+    logger:debug("ElementOut: ~p~n", [ElementOut]),
+    sign_signatures(Tail, ElementOut, PrivateKeyFun, CertBin);
+sign_signatures([], ElementIn, _, _) ->
+    ElementIn.
+
+get_child(Prefix, Name, #xmlElement{name = Pname, content = Content} = _Elem) ->
+    Qname = list_to_atom(Prefix ++ ":" ++ Name),
+    case [ E || #xmlElement{name = N} = E <- Content, N == Qname ] of
+        [] ->
+            error({no_such_child, Qname, Pname});
+        [Child] ->
+            Child
+    end.
+
+replace_child(#xmlElement{parents = [{Pname, _}|_], pos = N} = Elem,
+              #xmlElement{name = Pname, content = Content} = Parent) ->
+    Len = length(Content),
+    Pre = lists:sublist(Content, N-1),
+    Post = lists:sublist(Content, N+1, Len - N),
+    %%logger:debug("Pre: ~p~nPost: ~p~n", [Pre, Post]),
+    Parent#xmlElement{content = Pre ++ [Elem | Post]}.
+
+dreplace_child(#xmlElement{parents = Parents} = Elem, Root) ->
+    logger:debug("Parents: ~p~n", [Parents]),
+    do_dreplace_child(lists:reverse(Parents), Elem, Root).
+
+do_dreplace_child([{P, _}], #xmlElement{parents = [{P, _}|_], pos = N} = Elem,
+                  #xmlElement{name = P, content = Content} = Parent) ->
+    Len = length(Content),
+    Pre = lists:sublist(Content, N-1),
+    Post = lists:sublist(Content, N+1, Len - N),
+    logger:debug("PreLen: ~p~nPostLen: ~p~n", [length(Pre), length(Post)]),
+    Parent#xmlElement{content = Pre ++ [Elem | Post]};
```

Comment:
While I am a big fan of damp code and copy pasting, this looks like it would be more readable if it was just a call to `replace_child(Elem, Parent)`. `dreplace_child` is short for `deep_replace_child` is it not? So it is recursively replacing children. Or is it working on different data for different purposes compared to the `replace_child` function?